Add option to inflate unknown space

shrijitsingh99 opened the pull request:
Signed-off-by: Shrijit Singh <shrijitsingh99@gmail.com>

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1628 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | turtlebot simulation |

---

## Description of contribution in a few bullet points

* Added option to inflate unknown space
* Made some small optimizations and changes to make code cleaner
## Description of documentation updates required from your changes

* Added new parameter `inflate_around_unknown`, so need to add that to default configs and documentation page

SteveMacenski commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -288,7 +289,7 @@ InflationLayer::enqueue(
 
     // push the cell data onto the inflation list and mark
-    inflation_cells_[distance].push_back(CellData(index, mx, my, src_x, src_y));
+    inflation_cells_[distance].emplace_back(CellData(index, mx, my, src_x, src_y));
```
You can remove the `CellData` in this case if you're going to emplace_back.

SteveMacenski commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -79,11 +80,13 @@ InflationLayer::onInitialize()
   node_->get_parameter(name_ + "." + "enabled", enabled_);
   node_->get_parameter(name_ + "." + "inflation_radius", inflation_radius_);
   node_->get_parameter(name_ + "." + "cost_scaling_factor", cost_scaling_factor_);
   node_->get_parameter(name_ + "." + "inflate_unknown", inflate_unknown_);
```
What's this parameter? I haven't see this before: `inflate_unknown`. Did you test to see what this does / if that actually resolves the issue? 

I did not know myself about this parameter. It looks like we ported this from ROS1 but there's never been documentation about it in ROS1 or ROS2. This may be a hidden feature that actually fixes it.

shrijitsingh99 replied:
If set to true it allows the inflated area to extend into unknown space too else it gets cut of at unknown areas.

SteveMacenski replied:
got it, so its correct to say that do 2 different things.

shrijitsingh99 replied:
yeah, exactly

SteveMacenski commented:
https://github.com/ros-planning/navigation/pull/839

This may also be good to include if you're interested in the inflation layer. Porting this to ROS2 would speed things up quite a bit.

shrijitsingh99 commented:
> I don't understand a bunch of these changes. Please explain. They don't see to follow the speed up PR or relate to inflating unknown space.

Since I was already working on the speedup you mentioned, I added some more optimizations and in general just cleaned up some stuff to be inline with modern practices:
* Convert `cached_costs_` and `cached_distances_` to vectors, they allow for automatic memory management, no need for `new` and `delete`.
* Convert `cached_costs_` and `cached_distances_` to 1D vectors, to allow better locality of reference.
* Introduced explicit static casts in-place of implicit and C-Styled casts and removed unnecessary casts.
* Added `override` keyword to mark implicit overriding of virtual functions.
* Fixed a issue mentioned in the code comments regarding `-numeric_limit<double>::max()`

SteveMacenski commented on nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:
```diff
@@ -108,21 +102,21 @@ class InflationLayer : public Layer
     if (distance == 0) {
       cost = LETHAL_OBSTACLE;
-    } else if (distance * resolution_ <= inscribed_radius_) {
+    } else if (euclidean_distance <= inscribed_radius_) {
```
Please revert this, you do an extra copy in that change + the variable isn't later re-used.

SteveMacenski commented on nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:
```diff
@@ -133,11 +127,13 @@ class InflationLayer : public Layer
-    unsigned int dx = abs(mx - src_x);
-    unsigned int dy = abs(my - src_y);
-    return cached_distances_[dx][dy];
+    unsigned int dx = (mx > src_x) ? mx - src_x : src_x - mx;
```
This is nearly identical to the implementation of abs (https://code.woboq.org/userspace/glibc/stdlib/abs.c.html), please revert for readability.

shrijitsingh99 replied:
I did this as initially `distanceLookup(int mx, int my, int src_x, int src_y)` was this but where ever it was called its arguments were `unsigned int`:
https://github.com/ros-planning/navigation2/blob/d21cf3366b06e3a63014e411708c2cbae2ae3647/nav2_costmap_2d/plugins/inflation_layer.cpp#L235

So it had to make 2 casts one from unsigned int to int then back to unsigned int when assigning to `dx` and `dy`.
This way there are no casts.

SteveMacenski replied:
Ah ok, I can live with that

SteveMacenski commented on nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp:
```diff
@@ -148,16 +144,16 @@ class InflationLayer : public Layer
-    unsigned int dx = abs(mx - src_x);
-    unsigned int dy = abs(my - src_y);
-    return cached_costs_[dx][dy];
+    unsigned int dx = (mx > src_x) ? mx - src_x : src_x - mx;
```
This is nearly identical to the implementation of abs (https://code.woboq.org/userspace/glibc/stdlib/abs.c.html), please revert for readability.

SteveMacenski commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -61,14 +60,14 @@ InflationLayer::InflationLayer()
-  last_max_x_(std::numeric_limits<float>::max()),
-  last_max_y_(std::numeric_limits<float>::max())
+  cache_length_(0),
+  last_min_x_(std::numeric_limits<double>::lowest()),
```
lowest appears to just be -max  ( https://en.cppreference.com/w/cpp/types/numeric_limits). Not sure what issue you were referring to that asked for that to be changed. Can you clarify / link?

SteveMacenski commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -112,13 +115,11 @@ InflationLayer::updateBounds(
-    *max_x = std::numeric_limits<float>::max();
-    *max_y = std::numeric_limits<float>::max();
+
+    *min_x = std::numeric_limits<double>::lowest();
```
lowest appears to just be -max ( https://en.cppreference.com/w/cpp/types/numeric_limits). Not sure what issue you were referring to that asked for that to be changed. Can you clarify / link?

shrijitsingh99 replied:
https://github.com/ros-planning/navigation2/blob/ecc42fb3879d470590b9d9ef893c7565cee3b69b/nav2_costmap_2d/plugins/inflation_layer.cpp#L115-L117

shrijitsingh99 commented:
> There are a bunch of missing changes that Jorge implemented. At the bare minimum you're missing
> 
> https://github.com/ros-planning/navigation2/blob/d21cf3366b06e3a63014e411708c2cbae2ae3647/nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp#L171
> the change between a map to a vector, which I think accounts for the vast majority of the speed up.
> 
I have that port ready, haven't pushed it currently.

> I think you should simplify to a direct port before trying to do anything else. Some of the changes you made actually will make the code a little slower.
I mostly made changes related to recommended C+= syntax so shouldn't negatively affect performance.

Haven't benchmarked the `cached_costs_` and `cached_distances_` but should be slightly faster due to locality, worst case there is no performance impact. 

Might have missed some stuff causing it to be slower. If you think all the changes are unneeded, will revert back.

SteveMacenski commented:
I would make all the changes he made. Jorge's a smart guy and if those changes were robustly checked from months of running on Magazino's robots, I'd go with that and take his changes at face value.

SteveMacenski commented:
@shrijitsingh99 it fails some linters and the unit tests for the inflation layer. Please make sure to run these tests locally + run the stack with these change to debug. It looks like there's a real bug in there.

shrijitsingh99 commented:
Yeah sorry about that. Have made separate issue for a bug in the tests. Lets get that merged first before proceeding with this. Opened a separate PR since I don't think it is a good idea modifying the test as well as the module together can lead to bugs creeping in.

SteveMacenski commented:
@shrijitsingh99 are you running on ROS2 master source build or Eloquent binaries?

You'd see crustify appearing differently if you did Eloquent, there was a master / Foxy update. Its basically just:
```
auto yo = some_function_long_name(first_thing,
  second_thing);
```

goes to 

```
auto yo = some_function_long_name(
  first_thing, second_thing);
```

Any line extensions start arguments on a new line.

shrijitsingh99 commented:
> @shrijitsingh99 are you running on ROS2 master source build or Eloquent binaries?

 Eloquent binaries

> You'd see crustify appearing differently if you did Eloquent, there was a master / Foxy update.

Yup, pretty much sums the issue I was facing. Will switch to source build.

SteveMacenski commented:
@shrijitsingh99 you should rebase to get the new test

SteveMacenski commented:
@shrijitsingh99 can you verify you see a CPU drop? I want to make sure we got what we came for with this.

SteveMacenski commented:
@shivaang12 please review

SteveMacenski commented:
Inflation layer is actually one of the best covered files in the repo https://codecov.io/gh/ros-planning/navigation2/src/master/nav2_costmap_2d/plugins/inflation_layer.cpp

Though I think that is mostly from integration testing that wouldn't actually come up with any issues unless it was _really_ messed up since the robot an still navigate as long as its not in the middle of LETHALs. 

@shrijitsingh99 that is a good point though. One of the very few uncovered lines is this PR adds https://codecov.io/gh/ros-planning/navigation2/src/master/nav2_costmap_2d/plugins/inflation_layer.cpp#L239 it would be helpful to add a test for checking with this param on / off that it inflates or doesn't inflate unknown space. 

An easy test would be creating a 10x10 costmap of FREE and in the middle a single unknown cell. Check when off that its neighbors are still FREE, when on, checking that they're > FREE.

shrijitsingh99 commented:
Great, will add some tests for `inflate_unknown_` and `inflate_around_unknown_`.

SteveMacenski commented:
@shrijitsingh99 any update on this? We should try to square up existing things before adding new things. 5 unmerged PRs isn't as good as 3 merged ones :wink:

shrijitsingh99 commented:
> @shrijitsingh99 any update on this? We should try to square up existing things before adding new things. 5 unmerged PRs isn't as good as 3 merged ones 😉

I have started work on it should be done by tomorrow morning.

> You will also need to rebase this on master, a bunch of emergency updates went in over the weekend.

Will do that 👍

SteveMacenski commented on nav2_costmap_2d/test/integration/inflation_tests.cpp:
```diff
@@ -137,7 +138,8 @@ void TestNode::validatePointInflation(
 
         unsigned char expected_cost = ilayer->computeCost(dist);
-        ASSERT_TRUE(costmap->getCost(cell.x_, cell.y_) >= expected_cost);
+//        ASSERT_TRUE(costmap->getCost(cell.x_, cell.y_) >= expected_cost)
```
Why is this commented out?

shrijitsingh99 replied:
Was testing out something,, mistakenly commented out.
Fixed.

SteveMacenski commented:
@shrijitsingh99 did you mess with build settings or setup a CI on your own fork? That build error is not from our stuff and when I click on it it shows as part of your account/fork

shrijitsingh99 commented:
> @shrijitsingh99 did you mess with build settings or setup a CI on your own fork? That build error is not from our stuff and when I click on it it shows as part of your account/fork

Yeah, I was trying to set up the I on my own fork for testing purposes.

I didn't know it will switch to my personal CI for PR's too. How should I do go about resolving this?

SteveMacenski commented:
> I didn't know it will switch to my personal CI for PR's too. How should I do go about resolving this?

A PR can be thought of as "I have my own version of this [thing] and I want to generate a diff and apply that patch to your [thing]". When the MR is opened, its going to look at your end of things to decide what to trigger/do so it'll naturally run your configs. In fact, you can open a pull request from a completely unrelated repository into this one if you liked and applied some random patch. The GitHub interface doesn't make that easy, but Git will fully allow for it.

tl;dr, change the branch that run your CI stuff against to something other than `master` or `*-devel`

shrijitsingh99 commented:
> When the MR is opened, its going to look at your end of things to decide what to trigger/do so it'll naturally run your configs. 

I generally have been using Azure Pipelines, it allows to me run my own CI on my fork, and on the PR it runs the upstream repo's CI so I assumed here there was a similar thing.

Is this CircleCI specific or am I missing something?

PS: Have unlinked my own CI.

SteveMacenski commented:
@shrijitsingh99 don't bother, @AlexeyMerzlyakov should, he's been thinking about this problem. Merging this since that's not part of this PR.

SteveMacenski commented:
Actually,  @shrijitsingh99 can you do one thing on this, can you branch off this branch, rebase on master as of right now, and then resubmit a PR with some trivial change? (add a comment, whatever) 

I want to see if this is from some state issue. It can be a draft PR
